File: java-example/src/test/java/org/onflow/examples/java/getProtocolState/GetProtocolStateAccessAPIConnectorTest.java
package org.onflow.examples.java.getProtocolState;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.onflow.flow.common.test.FlowEmulatorProjectTest;
import org.onflow.flow.common.test.FlowTestClient;
import org.onflow.flow.sdk.FlowAccessApi;
import org.onflow.flow.sdk.FlowBlock;
import org.onflow.flow.sdk.FlowSnapshot;

import static org.junit.jupiter.api.Assertions.*;

@FlowEmulatorProjectTest(flowJsonLocation = "../flow/flow.json")
public class GetProtocolStateAccessAPIConnectorTest {
    @FlowTestClient
    private FlowAccessApi accessAPI;
    private FlowBlock block;

    private GetProtocolStateAccessAPIConnector protocolStateConnector;

    @BeforeEach
    public void setup() {
        protocolStateConnector = new GetProtocolStateAccessAPIConnector(accessAPI);
    }

    @Test
    public void canGetLatestProtocolStateSnapshot() {
        FlowSnapshot latestSnapshot = protocolStateConnector.getLatestProtocolStateSnapshot();
        assertNotNull(latestSnapshot, "Latest snapshot should not be null");
    }

    @Test
    public void canGetProtocolStateSnapshotByBlockId() {
        block = getLatestBlock();
        FlowSnapshot snapshot = protocolStateConnector.getProtocolStateSnapshotByBlockId(block.getId());
        assertNotNull(snapshot, "Snapshot should not be null");
    }

    @Test
    public void canGetProtocolStateSnapshotByHeight() {
        block = getLatestBlock();
        FlowSnapshot snapshot = protocolStateConnector.getProtocolStateSnapshotByHeight(block.getHeight());
        assertNotNull(snapshot, "Snapshot should not be null");
    }

    private FlowBlock getLatestBlock() {
        FlowAccessApi.AccessApiCallResponse<FlowBlock> response = accessAPI.getLatestBlock(true, false);
        if (response instanceof FlowAccessApi.AccessApiCallResponse.Success) {
            return ((FlowAccessApi.AccessApiCallResponse.Success<FlowBlock>) response).getData();
        } else {
            FlowAccessApi.AccessApiCallResponse.Error errorResponse = (FlowAccessApi.AccessApiCallResponse.Error) response;
            throw new RuntimeException(errorResponse.getMessage(), errorResponse.getThrowable());
        }
    }
}

File: kotlin-example/src/test/kotlin/org/onflow/examples/kotlin/getProtocolState/GetProtocolStateAccessAPIConnectorTest.kt
package org.onflow.examples.kotlin.getProtocolState

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.onflow.examples.kotlin.getTransaction.GetTransactionAccessAPIConnectorTest
import org.onflow.flow.common.test.FlowEmulatorProjectTest
import org.onflow.flow.common.test.FlowTestClient
import org.onflow.flow.sdk.FlowAccessApi
import org.onflow.flow.sdk.FlowBlock
import org.onflow.flow.sdk.FlowSnapshot

@FlowEmulatorProjectTest(flowJsonLocation = "../flow/flow.json")
internal class GetProtocolStateAccessAPIConnectorTest {
    @FlowTestClient
    lateinit var accessAPI: FlowAccessApi
    lateinit var block: FlowBlock

    private lateinit var protocolStateConnector: GetProtocolStateAccessAPIConnector

    @BeforeEach
    fun setup() {
        protocolStateConnector = GetProtocolStateAccessAPIConnector(accessAPI)
        block = GetTransactionAccessAPIConnectorTest.fetchLatestBlockWithRetries(accessAPI)
    }

    @Test
    fun `Can get latest protocol state snapshot`() {
        val latestSnapshot: FlowSnapshot = protocolStateConnector.getLatestProtocolStateSnapshot()
        assertNotNull(latestSnapshot, "Latest snapshot should not be null")
    }

    @Test
    fun `Can get protocol state snapshot by blockId`() {
        val latestSnapshot: FlowSnapshot = protocolStateConnector.getProtocolStateSnapshotByBlockId(block.id)
        assertNotNull(latestSnapshot, ("Snapshot should not be null"))
    }

    @Test
    fun `Can get protocol state snapshot by height`() {
        val latestSnapshot: FlowSnapshot = protocolStateConnector.getProtocolStateSnapshotByHeight(block.height)
        assertNotNull(latestSnapshot, ("Snapshot should not be null"))
    }
}
